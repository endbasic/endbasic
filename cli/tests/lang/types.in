' EndBASIC
' Copyright 2020 Julio Merino
'
' Licensed under the Apache License, Version 2.0 (the "License"); you may not
' use this file except in compliance with the License.  You may obtain a copy
' of the License at:
'
'     http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
' WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
' License for the specific language governing permissions and limitations
' under the License.

' Tests for types, including type annotations and type inference.
'
' This is must be run through the REPL (as opposed to running as a script)
' because the statements below may trigger compilation errors and abort
' execution right away.  The REPL compiles each statement separately so it
' can print those errors.

PRINT ">>> Type inference"
b = FALSE: b? = FALSE: PRINT b; b?
d = 0.0: d# = 0.0: PRINT d; d#
i = 3: i% = 3: PRINT i; i%
s = "Hello": s$ = "Hello": PRINT s; s$

PRINT ">>> Automatic type promotion"
d = 2.4: d = d + 3: PRINT d
i = 5: i = i + 2.6: PRINT i

PRINT ">>> Inconsistent type annotations in expressions"
d = 0.0: PRINT 2.1 + d? + 3.1
DIM a1(10) AS INTEGER: PRINT 3 + a1#(2) + 8

PRINT ">>> Inconsistent type annotations in commands"
i = 0: PRINT i#
DIM a2(10) AS INTEGER: PRINT a2#(2)

PRINT ">>> Type annotations in function calls"
PRINT LEN("foo")
PRINT LEN%("foo")
PRINT LEN$("foo")

PRINT ">>> Invalid type access"
i = 0: PRINT i(3)
PRINT LEN
PRINT PRINT

PRINT ">>> Argless function calls"
PRINT PI
PRINT PI()

PRINT ">>> Function references"
INPUT PI
