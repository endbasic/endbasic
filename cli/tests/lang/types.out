
    EndBASIC X.Y.Z
    Copyright YYYY-YYYY Julio Merino

    Type HELP for interactive usage information.

>>> Type inference
FALSE FALSE
 0  0
 3  3
HelloHello
>>> Automatic type promotion
 5.4
 8
>>> Inconsistent type annotations in expressions
ERROR: 1:22: Incompatible type annotation in d? reference
ERROR: 1:34: Incompatible type annotation in a1# reference
>>> Inconsistent type annotations in commands
ERROR: 1:14: Incompatible type annotation in i# reference
ERROR: 1:30: Incompatible type annotation in a2# reference
>>> Type annotations in function calls
 3
 3
ERROR: 1:7: Incompatible type annotation in LEN$ reference
>>> Invalid type access
ERROR: 1:14: i is not an array nor a function
ERROR: 1:7: In call to LEN: function requires arguments
ERROR: 1:7: PRINT is not an array nor a function
>>> Argless function calls
 3.141592653589793
ERROR: 1:7: In call to PI: expected no arguments nor parenthesis
End of input by CTRL-D
